### PR TITLE
persist: initial impl of S3Blob storage

### PR DESCRIPTION
**Pull request by danhhz**
### Motivation

  * This PR adds a feature that has not yet been specified.

### Description

S3 will be the primary production impl of persist Blob storage.

There are a number of key TODOs left as a followup. As a result, it's not hooked up to anything, but I've manually tested that this works end-to-end if I plug it into persistcfg.

### Checklist

- [x] This PR has adequate test coverage.
- [x] This PR adds a release note for any user-facing behavior changes.

### CODE REVIEW

**Review by aljoscha — COMMENTED** [review]
The code looks good, but S3 doesn't fit our (current) requirements well. I don't know how to fix that, off the top of my head (not surprising!). I could write some thoughts here but it's probably easier to brainstorm this in a call/chat.

**Review by aljoscha — APPROVED** [review]
I wasn't quite clear before. I think this is good to merge, despite the TODOs around consistency/racy-ness.

**Review by ruchirK — APPROVED** [review]
LGTM and very exciting!

File: src/persist/src/s3.rs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        let val = body.into_bytes().to_vec();
+        Ok(Some(val))
+    }
+
+    async fn set(&self, key: &str, value: Vec<u8>, allow_overwrite: bool) -> Result<(), Error> {
+        let client = self.ensure_open()?;
+        let path = self.get_path(key);
+
+        if !allow_overwrite {
+            // TODO: This is both inefficient and race-y, but it's unclear if
```
[the remark this answers is not in the document]
Comment (danhhz) [reply src/persist/src/s3.rs]: I think you're right but now that I think about it more, we've got to be careful around actual concurrent usage of Blob implementors (which we might want to allow once we move compaction out into its own thread). I added a note

File: src/persist/src/storage.rs [hunk 2]
```diff
@@ -79,7 +79,7 @@ pub trait Blob {
 
     /// Remove a key from the map.
     ///
-    /// Errors if the key does not exist.
+    /// Succeeds if the key does not exist.
```
[the remark this answers is not in the document]
Comment (danhhz) [reply src/persist/src/storage.rs:82]: Yup, theres no way to provide these semantics on s3 except by first doing a get, which is wasteful given that these semantics were mostly arbitrary in the first place. (I do, however, think it's valuable for all Blob impls to have the same semantics for this case and not an "undefined behavior" or "implementation defined" sort of situation.)

**Review by danhhz — COMMENTED** [review]
Updates this with a pretty major change to use the current rusoto library for s3 access. As discussed offline with @benesch and @quodlibetor, _someone_ needs to switch us to the new library, but it doesn't make sense for persist to be the trailblazer.

Also hopefully resolved how this should work with CI 🤞

**Comment by danhhz** [comment]
> Also hopefully resolved how this should work with CI 🤞

```
---- s3::tests::s3_blob stdout ----
--
  | Error: String("unavailable AWS_DEFAULT_REGION: environment variable not found")
```

Guess not, paging @benesch. Looks like the plugin ran and set env vars so not sure why my test didn't see them (AWS_DEFAULT_REGION is the first one it looks for). Something to do with mzcompose?

**Comment by danhhz** [comment]
Getting closer...

```
---- s3::tests::s3_blob stdout ----
--
  | Error: String("Error during dispatch: error trying to connect: error:1416F086:SSL routines:tls_process_server_certificate:certificate verify failed:ssl/statem/statem_clnt.c:1914: (unable to get local issuer certificate)")
```

**Comment by danhhz** [comment]
It looks like that last error is because rust-tls can't find system certs. I tried digging into how testdrive makes this work, but I'm not familiar enough with mzcompose and our testdrive configs to do it easily.

This feels like the type of thing that will take Nikhil 1 minute to fix, but me an hour. Given that it's easy to move the CI stuff into a followup, I'm going to do that to unblock some followup s3 work.

**Comment by danhhz** [comment]
Followup PR for CI bits https://github.com/MaterializeInc/materialize/pull/8444

**Pull request merged**